### PR TITLE
[WIP] gdrive: .ls()/.walk()/.walk_dir()

### LINKED ISSUE #5410: gdrive: _list_paths/walk_files returns both files and directories
This is a pre-requisite to properly handle `import-url`s and to generalize GdriveTree to the level of other trees so we could work with arbitrary fs tree structures. We should implement fsspec-like `ls()` and use it to implement `walk()` properly. Note that there is `_ids_cache` primitive caching that will definitely get in the way as it makes some strong assumptions about current location, so we might need to generalize it for the whole tree (kinda like fsspec's dircache).

https://github.com/intake/gdrivefs/blob/master/gdrivefs/core.py might be useful for inspiration.

### PR DESCRIPTION
**Pull request by isidentical**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏
Resolves #5410

### CODE REVIEW

File: dvc/tree/gdrive.py [hunk 1]
```diff
@@ -510,35 +511,68 @@ def exists(self, path_info, use_dvcignore=True):
...
+    ):  # pylint: disable=arguments-differ
+        if recursive:
+            for entry in self.walk_files(path_info, detail=detail):
+                if detail:
+                    yield entry
+                else:
+                    yield str(entry)
+            return None
+
+        parent_ids = self._path_to_item_ids(path_info.path)
```
Comment (shcheklein) [inline dvc/tree/gdrive.py:525]: minor: a bit confusing name here - they are just ids here, not parent ids as far as I understand

File: dvc/tree/gdrive.py [hunk 2]
```diff
@@ -510,35 +511,68 @@ def exists(self, path_info, use_dvcignore=True):
...
+                    yield {
+                        "type": "file",
+                        "name": item_path,
+                        "size": item["fileSize"],
+                        "checksum": item["md5Checksum"],
+                    }
+            else:
+                yield item_path
+
+        if parent_ids:
```
Comment (shcheklein) [inline dvc/tree/gdrive.py:550]: since we cache only `min()` next time this call can return a different result?

File: dvc/tree/gdrive.py [hunk 3]
```diff
@@ -510,35 +511,68 @@ def exists(self, path_info, use_dvcignore=True):
...
     def walk_files(self, path_info, **kwargs):
-        use_prefix = kwargs.pop("prefix", False)
-        if path_info == self.path_info or not use_prefix:
-            prefix = None
-        else:
-            prefix = path_info.path
-        for fname in self._list_paths(prefix=prefix, **kwargs):
-            yield path_info.replace(fname)
+        if kwargs.pop("prefix", False):
+            path_info = path_info / ""
```
Comment (shcheklein) [inline dvc/tree/gdrive.py:573]: hmm, not sure I understand what's going on here ... could you please clarify? is it equivalent? do we change upstream a bit to pass path/prefix in a different way?

File: dvc/tree/gdrive.py (hunk 2 above)
Comment (isidentical) [reply dvc/tree/gdrive.py:550]: No, not really. I can manually verify that caching works, and also this usage is something that I borrowed from the code above. Though I am not quite sure about why and how (probably the parent_ids is always 1-length list? maybe not but it seems that way)

File: dvc/tree/gdrive.py (hunk 3 above)
Comment (isidentical) [reply dvc/tree/gdrive.py:573]: I think this should be obsolete now

**Comment by isidentical** [comment]
GDrive tests doesn't work on my repo, so closing this PR in order to get an upstream branch

**Pull request closed without merging**